x/transferinject/types/packet_metadata.go

```diff
@@ -0,0 +1,57 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// MemoData represents the structure of the memo with user and hub metadata
+type MemoData struct {
+	UserMemo       string          `json:"user_memo"`
+	PacketMetadata *PacketMetadata `json:"packet_metadata"`
+}
```

Comment:
This is good but I have a suggestion that will bring us more inline with recommended practices, and be more maintainable

First of all, taking care of the user memo should be out of scope of this middleware
We should have an additional middleware (can be very lightweight, just 1 file) which always executes first, which takes the user memo and puts it in a json `{"user":<xxx>"}`

Then, your middleware can assume the memo is already json (map[string]any)
Then your middleware should simply add another key to this json

{
  "user": "foobar",
  "transferinject" : {
    "denom_metadata" : ...
  },
  "other_module" : ...
}


I feel quite strongly about this change tbh, especially because I will already add another memo type in the work I'm doing
And it's not unlikely we will add more later

This is also best practice according to https://github.com/cosmos/ibc/tree/main/spec/app/ics-020-fungible-token-transfer#using-the-memo-field